// web3-provider/dist/lib/Web3Provider.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const web3_1 = __importDefault(require("web3"));
const ProxiedProperty_1 = __importDefault(require("./ProxiedProperty"));
const defaultProviderOptions = {
    retriesOnError: 2,
    retryOnNullResponse: false
};
class Web3Provider {
    constructor(web3Endpoints, options) {
        this.instances = web3Endpoints.map(p => new web3_1.default(p));
        this.currentInstanceIndex = 0;
        this.options = Object.assign({}, defaultProviderOptions, options);
        return new Proxy(this, { get: this.get });
    }
    get(target, prop) {
        if (target.hasOwnProperty(prop))
            return target[prop];
        target.currentInstanceIndex = (target.currentInstanceIndex + 1) % target.instances.length;
        const currentInstance = target.instances[target.currentInstanceIndex];
        if (currentInstance[prop] === undefined)
            return undefined;
        return new ProxiedProperty_1.default(currentInstance[prop], target, [{ name: prop }]);
    }
    nextInstance() {
        this.currentInstanceIndex = (this.currentInstanceIndex + 1) % this.instances.length;
        const currentInstance = this.instances[this.currentInstanceIndex];
        return currentInstance;
    }
}
exports.default = Web3Provider;
